lib/memjs/memjs.js

```diff
@@ -170,6 +186,137 @@ Client.prototype.get = function(key, callback) {
+    var key = keys[keyIdx];
+    bytesWritten += addToRequestBuffer(constants.OP_GETKQ, key, '', '', this.seq, request, bytesWritten);
+  }
+
+  bytesWritten += addToRequestBuffer(
+    constants.OP_NO_OP, '', '', '', this.seq, request, bytesWritten);
+
+  return request;
+};
+
+/** Executing a pipelined (multi) get against a single server. This is a private implementation detail of getMulti. */
+Client.prototype._getMultiToServer = function (serv, keys, callback) {
+  var self = this;
+
+  var responseMap = {};
+
+  var handle = function(response) {
+    switch (response.header.status) {
+    case  0:
+      if (callback) {
+        var deserialized = self.serializer.deserialize(response.header.opcode, response.val, response.extras);
+        // When we get the no-op response, we are done with this one getMulti in the per-backend fan-out
+        if (response.header.opcode === constants.OP_NO_OP) {
+          // this ensures the handler will be deleted from the responseCallbacks map in server.js
+          handle.quiet = true;
+          callback(null, responseMap);
+        } else {
+          var key = response.key.toString();
+          responseMap[key] = deserialized.value;
+        }
+      }
+      break;
+    case 1:
+      if (callback) { callback(null, null, null); }
+      break;
+    default:
+      var errorMessage = 'MemJS GET: ' + errors[response.header.status];
+      self.options.logger.log(errorMessage);
+      if (callback) { callback(new Error(errorMessage), null, null); }
+    }
+  };
+  // This prevents the handler from being deleted
+  // after the first response. Logic in server.js.
+  handle.quiet = true;
+
+  var request = this._buildGetMultiRequest(keys);
+  serv.onResponse(this.seq, handle);
+  this.incrSeq();
+  serv.onError(this.seq, function(err) {
+    if (callback) { callback(err, serv.host + ':' + serv.port, null); }
+  });
+  serv.write(request);
+};
+
+// MULTI-GET / GET-MULTI
+//
+// Retrieves the value at the given keys in memcache.
+//
+// The callback signature is:
+//
+//     callback(err, values, flags)
+//
+// _values_ is a map of key->string
+// _flags_ is a `string`.
+Client.prototype.getMulti = function(keys, callback) {
+  var self = this;
+  if(callback === undefined) {
+    return promisify(function(callback) {
+      self.getMulti(keys, function(err, values, flags) {
+        callback(err, {values: values, flags: flags});
+      });
+    });
+  }
+
+
+  var serverIndexToKeys = {};
+  keys.forEach(function(key)  {
+    var serverIndex = self.serverIndex(key);
+    if (!serverIndexToKeys[serverIndex]) {
+      serverIndexToKeys[serverIndex] = [];
+    }
+    serverIndexToKeys[serverIndex].push(key);
+  });
+
+  var outstandingCalls = Object.keys(serverIndexToKeys).length;
+  var recordMap = {};
+  var hadError = false;
+  function latchCallback(err, values, flags) {
+    if (hadError) {
+      return;
+    }
+
+    if (err) {
+      hadError = true;
+      callback(err, null, null);
```

Comment:
what cases does `err` cover? Is it connection errors, parsing errors, or does it include the server replying "hey we don't have this key"?
a) comment here might be good about this
b) could we ever want to return a partially filled `recordMap`?

Reply:
will add comment

err is just thrown errors - missing key is *not* an error (also, we are using "getkq - get key quiet" which doesn't even reply on missing key)

re: returning a partially filled recordMap - thought about this one some and ... I think no? it seems really unpredictable if you're calling five memcache backends and one goes bad, maybe the bad one replies first and this is empty, or the bad one replies last and it's mostly full, would the client really want to deal with that ... hmmm ... I mean, there's no harm in passing the recordMap back here and the client can choose if they want to use it? okay fine, yes

Reply:
Well I think you’re right that it’s weird to return inconsistently complete stuff...